mail/src/main/java/org/jboss/as/mail/extension/MailSessionService.java

```diff
@@ -22,48 +22,47 @@
 
 package org.jboss.as.mail.extension;
 
-import java.util.Map;
+import java.util.function.Consumer;
 import java.util.function.Supplier;
 
-import org.jboss.as.network.OutboundSocketBinding;
-import org.jboss.msc.service.Service;
+import jakarta.mail.Session;
+import jakarta.mail.util.StreamProvider;
+
+import org.jboss.msc.Service;
 import org.jboss.msc.service.StartContext;
 import org.jboss.msc.service.StartException;
 import org.jboss.msc.service.StopContext;
-
-import jakarta.mail.Session;
+import org.wildfly.security.manager.WildFlySecurityManager;
 
 /**
  * Service that provides a jakarta.mail.Session.
- *
+ * TODO Consider removing this. It has no consumers.
  * @author Tomaz Cerar
  * @created 27.7.11 0:14
  */
-public class MailSessionService implements Service<Session> {
-    private final MailSessionConfig config;
-    private final Map<String, Supplier<OutboundSocketBinding>> socketBindings;
-    private volatile SessionProvider provider;
+public class MailSessionService implements Service {
+    private final Consumer<Session> session;
+    private final Supplier<SessionProvider> provider;
 
-    public MailSessionService(final MailSessionConfig config, final Map<String, Supplier<OutboundSocketBinding>> socketBindings) {
-        MailLogger.ROOT_LOGGER.tracef("service constructed with config: %s", config);
-        this.config = config;
-        this.socketBindings = socketBindings;
-    }
-
-    MailSessionConfig getConfig() {
-        return config;
+    public MailSessionService(Consumer<Session> session, Supplier<SessionProvider> provider) {
+        this.session = session;
+        this.provider = provider;
     }
 
+    @Override
     public void start(final StartContext startContext) throws StartException {
-        MailLogger.ROOT_LOGGER.trace("start...");
-        provider = SessionProviderFactory.create(config, socketBindings);
+        // StreamProvider.provider() loads via TCCL.
+        ClassLoader loader = WildFlySecurityManager.getCurrentContextClassLoaderPrivileged();
+        try {
+            WildFlySecurityManager.setCurrentContextClassLoaderPrivileged(StreamProvider.class);
+            this.session.accept(this.provider.get().getSession());
+        } finally {
+            WildFlySecurityManager.setCurrentContextClassLoaderPrivileged(loader);
+        }
```

Comment:
MSC threads do not contain StreamProvider implementations in their context class loader.
Unless we manipulate the TCCL in this fashion, this service would otherwise fail to start 100% of the time.

Reply:
My understanding is the Mail session will be retrieved via the `this.provider.get().getSession()`. I'm suggesting here to let the responsibility to configure the TCCL to load the Mail Session to the SessionProvider implementation.

We are already done it at [ManagedSession](https://github.com/wildfly/wildfly/blob/main/mail/src/main/java/org/jboss/as/mail/extension/SessionProviderFactory.java#L172) and we will need to do it for the [DirectSessionProvider](https://github.com/wildfly/wildfly/blob/main/mail/src/main/java/org/jboss/as/mail/extension/SessionProviderFactory.java#L259) as well in a follow up Jira to allow CDI injection, since I think right now we cannot inject Mail Sessions using CDI when the sessions are created by using the mail annotations.

Does it make sense to you?

Reply:
@yersan OK - I understand now.  I see that you fixed the TCCL issue within ManagedSession as part of the recently merged WFLY-17470.  Addressing this within the SessionProvider implementations makes sense.
I'll drop this commit and rebase on main.

Reply:
Perfect! thanks @pferraro, LGTM